Issue 1498 - Adding entity fields

gkovats opened the pull request:
- Closes #1498

Adds Entity form fields to Settings page:
- 5 fields: entity, entity_email, entity_url, entity_phone, entity_location
- limits all to 100 characters: wanted to limit because this will add weight to every meeting in every feed.
  - *Note*: the added weight is worth bringing up to Meeting Guide, they'll likely notice
- added defaults for entity, entity_url, entity_email, so we start with values from WordPress

Adds Entity fields to export / import
- sanitizer caps strings off at 100 characters
- export supplies the local entity values for meetings without a data_source, otherwise grabs any entity values from post_meta / meeting fields

Adds Entity fields to TSML Legacy
- Added in "Updated" box. I took a crack at it, seemed like it belonged their, but I know we want to be efficient with screen real estate, and it should follow the pattern that TSML UI and Meeting Guide will use. Consider this a draft.

![image](https://github.com/user-attachments/assets/98881e12-eb28-46ab-9be7-95a007a3dcb3)

gkovats commented on includes/admin_settings.php:
```diff
@@ -332,7 +346,59 @@ function tsml_settings_page()
                         </form>
                     </div>
 
```
Check the order of these on the settings screen. I ended up putting this in spot 2, left column, and moved the Feeds settings to top of the middle column. But I don't stare at this screen much, so it's a guess.

gkovats commented on includes/functions.php:
```diff
@@ -1774,6 +1791,11 @@ function tsml_sanitize_import_meetings($meetings, $data_source_url = null, $data
         }
 
+        //cleanup entity fields
+        foreach($tsml_entity_fields as $field) {
```
I can defend this, but agree it looks awful.

joshreisner commented on includes/admin_settings.php:
```diff
@@ -332,7 +347,56 @@ function tsml_settings_page()
+                                <?php _e('Service Entity Information', '12-step-meeting-list') ?>
+                            </h2>
+                            <p>
+                                <?php echo sprintf(__('Enter information for your service entity here to help identity the meeting source when sharing feeds with others.', '12-step-meeting-list'), get_post_type_archive_link('tsml_meeting')) ?>
```
don't need the `sprintf` here - we'd only need that if you were adding params to the i18n string

joshreisner commented on includes/admin_settings.php:
```diff
@@ -202,6 +202,21 @@ function tsml_settings_page()
+            }
+            update_option('tsml_entity', $tsml_entity);
+            tsml_cache_rebuild(); //these values affects what's in the cache
+        }
```
i think we want a `tsml_alert` here to say that the update was successful

joshreisner commented on templates/single-meetings.php:
```diff
@@ -359,8 +359,35 @@ class="btn btn-default btn-block contact-phone">
+                                        <?php the_modified_date() ?>
+                                    </p>
+                                    <?php
+                                    if (!empty($meeting->{'data_source'})) {
```
i think we need some intro text here potentially like "This meeting listing is provided by:"

let's show the email address - this is more for administrative purposes

the feedback email form should either be hidden and we add something here, or it should send to the entity's feedback_email addresses if present

minor: could we use the `$meeting->data_source` pattern for this and the other vars?

minor: i would like to steer away from using `<br>` tags - could this be Ps or a UL?

gkovats replied:
Adding "This meeting listing is provided by:" only concern it will show up everywhere if we display for local _and_ imported meetings, and it just adds that one more line to what your eye is scanning for
![image](https://github.com/user-attachments/assets/181a9d0c-1da1-4bce-905a-7f3bddd4e221)

Look correct to you? Should it be smaller?

gkovats replied:
> or it should send to the entity's feedback_email addresses if present

Honestly this raises questions I want to talk about, cause it'll be a mess typing it out. Don't feel equipped with enough history on this tool to answer them myself.

joshreisner replied:
hm, yeah, i'm not so sure. perhaps less is more?

my reason for wanting to add it is that we want to make it clear what this section is - some users might get confused and think it is meeting information.

the purpose of this section is to be like this section in the app:
![Screenshot 2024-09-24 at 3 11 42 PM](https://github.com/user-attachments/assets/1c4dcfb7-dc2d-4812-ad8c-c4bcf47fba76)

if we could replicate the headline style for the other sections for the entity name, and make the url and phone buttons like in this pic:

<img width="330" alt="Screenshot 2024-09-24 at 3 14 05 PM" src="https://github.com/user-attachments/assets/0ea076c4-31b8-4acb-ada6-19a98f40154a">

we might try the `This meeting listing is provided by:` above the headline in a `<small>`?

joshreisner replied:
here's what it currently looks like in [this PR for TSML UI](https://github.com/code4recovery/tsml-ui/pull/416):

![image](https://github.com/user-attachments/assets/eb5fd843-cd10-4611-adfb-9ee60df1122e)

not that it needs to match per se but just for context

joshreisner commented on includes/admin_settings.php:
```diff
@@ -332,7 +347,56 @@ function tsml_settings_page()
+
+                            <form method="post" class="stack compact" action="<?php echo $_SERVER['REQUEST_URI'] ?>">
+                                <?php wp_nonce_field($tsml_nonce, 'tsml_nonce', false) ?>
+                                <h3>
```
could these `h3`s be `label`s instead? the smaller type should look good as well

i would keep it `Entity Name` but change:

`Entity Contact Email` -> `Administrative Contact Email`
`Entity Contact Phone` -> `Public Phone Number`
`Entity Location` -> `Your Service Area`
`Entity Website` -> `Website Address`

gkovats replied:
"Your" felt redundant on looking at it, like it could be in all labels. This OK, or you think the "your" belongs?
![image](https://github.com/user-attachments/assets/7cb69766-a1c1-4b6b-8d8d-d2471debda77)

joshreisner replied:
that's good! maybe make them `label`s though, if they're not already

joshreisner commented on includes/variables.php:
```diff
@@ -76,6 +76,15 @@
+    'entity_phone',
+    'entity_location',
+    'entity_url',
+];
```
do we need to add feedback email here? the idea is that they should pass through from the imported service entity

joshreisner reviewed (commented):
look great, thanks for jumping on this so quickly! just a few comments

gkovats commented on includes/functions.php:
```diff
@@ -1774,6 +1790,11 @@ function tsml_sanitize_import_meetings($meetings, $data_source_url = null, $data
 
+        //cleanup entity fields
+        foreach($tsml_entity_fields as $field) {
+            $meetings[$i][$field] = substr(trim(strval(isset($meetings[$i][$field]) ? $meetings[$i][$field] : '')), 0, 100);
```
Thanks, updated.

joshreisner commented:
one minor thing: i'm seeing PHP warnings on some of the entity form fields when running locally:

<img width="382" alt="Screenshot 2024-10-12 at 8 59 23 AM" src="https://github.com/user-attachments/assets/0f23bf2e-4322-4789-b53c-86472d6a7830">

joshreisner commented:
also i think `feedback_emails` should use `array_values` - see example at https://demo.code4recovery.org/wp-admin/admin-ajax.php?action=meetings

<img width="270" alt="Screenshot 2024-10-12 at 9 07 16 AM" src="https://github.com/user-attachments/assets/4936b1fe-8cd5-4100-af43-0473eedd257b">

joshreisner commented:
one last piece of feedback, and feel free to push back and/or say we should make this a subsequent PR, is i think if feedback emails are present on a meeting in legacy, we should expose the feedback form and send to those emails.

ie here:
```
                        if (!empty($tsml_feedback_addresses)) { ?>
```
we might say 
```
                        if (!empty($tsml_feedback_addresses) || !empty($meeting->feedback_emails) { ?>
```

gkovats commented:
@joshreisner  FYI
- added `array_values()` to places where I'm casting to array to avoid keys
- fixed the admin settings form to ensure those array fields are present before displaying
- updated form button to use already existing button style class `btn btn-default btn-block` and removed my new style
- fixed the logic for displaying the change request form - it was entirely an oversight on my part

joshreisner commented:
looks awesome! only detail i noticed is the link icon is a little big - looks like it just needs width and height props

if you don't mind i'll just go ahead and merge this add them in the follow-on PR where i set the new version

<img width="315" alt="Screenshot 2024-10-13 at 7 41 02 PM" src="https://github.com/user-attachments/assets/298ac22d-f8ca-4c34-9bb1-b237ea7429ef">
